allennlp/training/trainer.py

```diff
@@ -348,7 +348,10 @@ def _enable_gradient_clipping(self) -> None:
         if self._grad_clipping is not None:
             # Pylint is unable to tell that we're in the case that _grad_clipping is not None...
             # pylint: disable=invalid-unary-operand-type
-            clip_function = lambda grad: grad.clamp(-self._grad_clipping, self._grad_clipping)
+            def clip_function(grad):
+                # Sparse tensors do not have a clamp method.
```

Comment:
Is there no way to get the same functionality with sparse tensors?  I feel like this should at least emit a warning if it can't do what you want, instead of silently failing.

Reply:
You should be able to clamp the values directly and create a new sparse tensor to return.

Reply:
True, but do I need to be careful about coalescing the sparse tensor before clamping the values?

Reply:
I don't think so.  See `sparse_clip_norm` for an example that clips gradient norms of sparse gradients.

Reply:
Thanks!